File: src/toolchain.py
import json
from typing import Dict, List, Iterator, Optional
from errors import ConfigError

class Step:
    def __init__(self, **kwargs):
        self.name           = kwargs.get('stepName', None)
        self.command        = kwargs.get('command', None)
        self.arguments      = kwargs.get('arguments', None)
        self.output         = kwargs.get('output', None)
        self.allow_error    = kwargs.get('allowError', False)
        self.uses_ins       = kwargs.get('usesInStr', False)
        self.uses_runtime   = kwargs.get('usesRuntime', False)
    
    def verify(self) -> Optional[ConfigError]:
        return None

    def to_dict(self) -> Dict:
        return {
            'stepName': self.name,
            'command': self.command,
            'arguments': self.arguments,
            'output': self.output,
            'allowError': self.allow_error,
            'usesInStr': self.uses_ins,
            'usesRuntime': self.uses_runtime
        }

    def __repr__(self):
        return json.dumps(self.to_dict(), indent=2)

class ToolChain:
    def __init__(self, name: str, steps: List[Step]):
        self.name = name
        self.steps = steps
    
    def verify(self) -> Optional[ConfigError]:
        errors = ConfigError()
        for step in self.steps:
            if error := step.verify():
                errors.add(error)
        return errors if errors else None
    
    def to_dict(self) -> Dict[str, List[Dict]]:
        return {self.name: [step.to_dict() for step in self.steps]}

    def __repr__(self):
        return json.dumps(self.to_dict(), indent=2)
    
    def __iter__(self) -> Iterator[Step]:
        return iter(self.steps)

    def __len__(self) -> int:
        return len(self.steps)

    def __getitem__(self, index: int) -> Step:
        return self.steps[index]

File: src/toolchain.rs
use std::path::Path;

use serde::Deserialize;

use crate::config::Executable;
use crate::error::{DragonError, Validate};

/// A single step in a toolchain (e.g., compile, link, run).
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Step {
    #[serde(rename = "exe", default)]
    pub exe_raw: String,
    #[serde(default)]
    pub args: Vec<String>,
    #[serde(default)]
    pub allow_error: bool,
    #[serde(rename = "usesInStr", default)]
    pub uses_ins: bool,
    #[serde(default)]
    pub uses_runtime: bool,
}

impl Step {
    /// Derive a human-readable step name from the raw exe string and the executable.
    pub fn display_name(&self, exe: &Executable) -> String {
        match self.exe_raw.as_str() {
            "$EXE" => exe.id.clone(),
            "$INPUT" => "run".to_string(),
            other => {
                // Use filename component for paths, bare name as-is
                Path::new(other)
                    .file_name()
                    .unwrap_or(other.as_ref())
                    .to_string_lossy()
                    .into_owned()
            }
        }
    }
}

impl Validate for Step {
    fn validate(&self) -> Vec<DragonError> {
        let mut errors = Vec::new();
        if self.exe_raw.is_empty() {
            errors.push(DragonError::Config(
                "Missing required field 'exe' in Step".into(),
            ));
        } else if !self.exe_raw.starts_with('$') && self.exe_raw.contains('/') {
            // Only check existence for paths (containing /), not bare names resolved via $PATH
            if !Path::new(&self.exe_raw).exists() {
                errors.push(DragonError::Config(format!(
                    "Cannot find exe '{}' in Step",
                    self.exe_raw
                )));
            }
        }
        errors
    }
}

/// An ordered sequence of Steps that form a compilation/execution pipeline.
#[derive(Debug, Clone)]
pub struct ToolChain {
    pub name: String,
    pub steps: Vec<Step>,
}

impl ToolChain {
    pub fn new(name: &str, steps: Vec<Step>) -> Self {
        Self { name: name.into(), steps }
    }

    pub fn len(&self) -> usize {
        self.steps.len()
    }

    pub fn is_empty(&self) -> bool {
        self.steps.is_empty()
    }

    pub fn iter(&self) -> std::slice::Iter<'_, Step> {
        self.steps.iter()
    }
}

impl Validate for ToolChain {
    fn validate(&self) -> Vec<DragonError> {
        self.steps.iter().flat_map(|s| s.validate()).collect()
    }
}
